ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs

```diff
@@ -99,62 +98,71 @@ type family HasResponder (appType :: AppType) :: Bool where
 --   serving downstream peers using server side of each protocol and getting
 --   updates from upstream peers using client side of each of the protocols.
 --
-data MuxApplication (appType :: AppType) ptcl m  where
+-- * the monadic action which wraps the callbacks will be run only once by the
+--   multiplexing layer.  This makes it safe for sharing a common mutable state
+--   between various protocol.
+--
+data MuxApplication (appType :: AppType) ptcl m bytes a where
   MuxInitiatorApplication
     -- Initiator application; most simple application will be @'runPeer'@ or
     -- @'runPipelinedPeer'@ supplied with a codec and a @'Peer'@ for each
     -- @ptcl@.  But it allows to handle resources if just application of
     -- @'runPeer'@ is not enough.  It will be run as @'ModeInitiator'@.
-    :: (ptcl -> Channel m ByteString ->  m ())
-    -> MuxApplication InitiatorApp ptcl m
+    :: m (ptcl -> Channel m bytes ->  m a)
+    -> MuxApplication InitiatorApp ptcl m bytes (a, Void)
```

Comment:
The use of `(a, Void)` here requires a comment. Why not instead have 2 parameters: one for the initiator result type, one for that of the responder?

Reply:
You mean `MuxApplication appType ptcl m a b`, right this could be fixed.  I tried this at first but I run into some problems, but I don't see why in the first place.